src/replication.c

```diff
@@ -157,15 +157,28 @@ replicator(void *arg) {
 		MTX_UNLOCK(&spec->rcommonq_mtx);
 
 		//Enqueue to individual replica cmd queues and send on the respective fds
-		read_cmd_sent = false;
+		if(cmd->opcode == ZVOL_OPCODE_READ)
+			cmd_read = true;
+		else
+			cmd_read = false;
 		MTX_LOCK(&spec->rq_mtx);
+		if(spec->healthy_rcount == 0) {
+			send_multiple_read = cmd->send_multiple_read = true;
+			healthy_untried = 0;
+		} else {
+			send_multiple_read = cmd->send_multiple_read = false;
+			healthy_untried = spec->healthy_rcount;
+		}
 
 		TAILQ_FOREACH(replica, &spec->rq, r_next) {
 			if(replica == NULL) {
 				REPLICA_LOG("Replica not present");
 				MTX_UNLOCK(&spec->rq_mtx);
 				exit(EXIT_FAILURE);
 			}
+			if(cmd_read && !send_multiple_read && replica->state != ZVOL_STATUS_HEALTHY) {
```

Comment:
replica lock required here..

Reply:
it should be ok.. not required a lock here as long as replica data structure is intact